include/jemalloc/internal/arena_structs.h

```diff
@@ -46,6 +46,11 @@ struct arena_s {
 
 	/* Synchronization: internal. */
 	arena_stats_t		stats;
+	/*
+	 * The current size of stats is 10384B.  To avoid false sharing,
+	 * pad it to the multiple of 128B.
+	 */
+	char padding[112];
```

Comment:
>  which in turn causes the fields below to be not cacheline aligned, and the slowdown we observed -- is that correct?

Yes.

Regarding the double check:

1. adding `char padding[16]` can reproduce the same slowdown.
2. As discussed offline: yes, with the new arena_statas_t, `a0->all_bins[0].lock.locked` has address `0x7ffff6014540` and `a0->all_bins[0].lock.prof_data.n_lock_ops` has address `0x7ffff6014510`. Thus, they are not on the same cacheline, causing more cache misses and therefore the slowdown.

Reply:
@guangli-dai does move the lock hint: https://github.com/jemalloc/jemalloc/blob/756d4df2fd1b5bde025abed50c9b771376d72c6f/include/jemalloc/internal/mutex.h#L53

to between lock and prof_data also resolve the benchmark issue?

I'm considering this because, sizeof(bin_t) is 224 which is not a multiple of cacheline as well, so even if we make sure bin[0] starts aligned, there will be bins not aligned still. Making sure the data touched as little spread as possible would be the right solution -- looking into pthread_mutex_t
```
lock.__data 
$15 = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 512, __spins = 0, 
  __elision = 0, __list = {__prev = 0x0, __next = 0x0}}
```
It looks on the common / non-contended case, the last 16 bytes at least won't be accessed -- therefore moving the locked flag on our side to the beginning will group all the common-case data together.

Reply:
It makes sense and can also properly resolve the slowdown on the tested benchmark. Included in this PR. 

Another question: Currently, the size of `malloc_mutex_t` is 112B. Should we consider align it with CACHELINE or make a sub-struct that is always cacheline aligned?

Reply:
> Should we consider align it with CACHELINE or make a sub-struct that is always cacheline aligned?

That's a good question. Aligning every bin / mutex would make sure each lock resides on exactly 1 cacheline, which is a pretty nice guarantee. However it also means every bin will include some padding byte, increasing the total footprint of hot cachelines used by the bins. We can benchmark that change and see if it helps, but my intuition is that might not help much, since in prod workload most of the bins (specially the small sizes) are hot enough to stay cached. So that w/o padding, the lock crossing cacheline boundaries isn't penalized much, and w/ padding the extra line might loss some efficiency too.

Reply:
I see. The size increase caused by padding is indeed a concern. Will do some more experiments in prod to explore this.